specs/DefaultDownloadDialog.md

```diff
@@ -0,0 +1,221 @@
+
+
+```c#
+[uuid(9139c04d-8f37-42ae-8b63-01940c34d22f), object, pointer_default(unique)]
+interface ICoreWebView2_6 : ICoreWebView2_5
+{
+  /// `TRUE` if the default download dialog is currently visible.
+  [propget] HRESULT IsDefaultDownloadDialogVisible(
+      [out, retval] BOOL* isDefaultDownloadDialogVisible);
+
+  /// Sets the `IsDefaultDownloadDialogVisible` property. A download is
+  /// considered recent if it was downloaded in this instance of the webview.
+  /// If there has not yet been a recent download, setting the default download
+  /// dialog to visible displays all past downloads for the current profile.
+  /// Otherwise, the dialog will display only the recent downloads with a
+  /// "See more" button, which displays the past downloads when clicked on.
+  ///
+  /// \snippet ViewComponent.cpp ToggleDefaultDownloadDialog
+  [propput] HRESULT IsDefaultDownloadDialogVisible(
+      [in] BOOL isDefaultDownloadDialogVisible);
+
+  /// Adds an event handler for the `DefaultDownloadDialogLayout` event. The event is
+  /// raised when the default download dialog's position needs to be recomputed,
+  /// such as when the dialog is first opened, when the webview is resized, and
+  /// when the webview is navigated. Use the `Position` property on the event
+  /// args to specify the new position. To set the position relative to the
+  /// current webview bounds, use the `LocalBounds` property on the event args.
+  ///
+  /// \snippet ViewComponent.cpp SetDefaultDownloadDialogPosition
+  HRESULT add_DefaultDownloadDialogLayout(
+      [in] ICoreWebView2DefaultDownloadDialogLayoutEventHandler* eventHandler,
+      [out] EventRegistrationToken* token);
+
+  /// Remove an event handler previously added with `add_DefaultDownloadDialogLayout`.
+  HRESULT remove_DefaultDownloadDialogLayout([in] EventRegistrationToken token);
+}
+
+/// Implements the interface to receive `DefaultDownloadDialogLayout` events.
+[uuid(86557854-794a-414a-b046-ba515f617306), object, pointer_default(unique)]
+interface ICoreWebView2DefaultDownloadDialogLayoutEventHandler : IUnknown {
+
+  /// Provides the event args for the corresponding event.
+  HRESULT Invoke([in] ICoreWebView2* sender,
+                 [in] ICoreWebView2DefaultDownloadDialogLayoutEventArgs* args);
+}
+
+[uuid(e7d2bdea-2e67-4550-b286-6e7d20653d4e), object, pointer_default(unique)]
+interface ICoreWebView2DefaultDownloadDialogLayoutEventArgs : IUnknown {
+  /// The default download dialog position of the top-left corner, relative to
+  /// the webview bounds. Use `LocalBounds` to check the current webview bounds.
+  /// The default position of the download dialog is calculated with the
+  /// following formula:
+  /// (LocalBounds.right - Width - 30, LocalBounds.top + 15).
+  [propget] HRESULT Position([out, retval] POINT* position);
+
+  /// Sets the default download dialog position. Returns `E_INVALIDARG` if an
+  /// invalid position is provided, and the default position is used instead.
```

Comment:
What would be an invalid position? Are you allowed to position the dialog outside or partially outside of the bounds of the webview2 (probably not?) Anything else considered invalid?

Reply:
An invalid position for the top-left corner of the dialog would be anything outside of the webview bounds, which means that the dialog can be partially outside of the webview bounds (this is the current behavior if the webview is small enough)

Reply:
In the sample code, if the WebView2 is small and the download dialog large, it is possible for both x_from_right and y_from_top to be negative. If the top-left corner needs to be within the WV2 bounds, then it sounds like the sample code needs to handle this scenario better.

This is where I think having an enum would be beneficial. For a small WebView2, forcing the download dialog to be inside the WV2 bounds could mean the download dialog covers most or all of the WebView content. With the enum, we could have internal metrics based on size of download dialog and WV2 to decide whether to place the dialog completely inside the WV2 bounds or to place it outside the bounds.

Reply:
Good point. I think the enum design would have to come with an offset, as the main reason we're providing this API is to ensure the dialog doesn't cover important parts of an application (ex: request was to lower default download dialog by x pixels). It might be weird to take an offset and then disregard it for small enough bounds?